api/provider/aws/aws_test/load_balancer_create_test.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/quintilesims/layer0/common/config/mock_config"
+	"github.com/quintilesims/layer0/common/models"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLoadBalancerCreate(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+	mockConfig.EXPECT().VPC().Return("vpc_id").AnyTimes()
+	mockConfig.EXPECT().Region().Return("region").AnyTimes()
+	mockConfig.EXPECT().AccountID().Return("123456789012")
+	mockConfig.EXPECT().PrivateSubnets().Return([]string{"priv1", "priv2"}).AnyTimes()
+	mockConfig.EXPECT().PublicSubnets().Return([]string{"pub1", "pub2"}).AnyTimes()
+
+	defer provider.SetEntityIDGenerator("lb_id")()
+
+	req := models.CreateLoadBalancerRequest{
+		LoadBalancerName: "lb_name",
+		EnvironmentID:    "env_id",
+		IsPublic:         true,
+		Ports: []models.Port{
+			{
+				CertificateName: "cert",
+				ContainerPort:   88,
+				HostPort:        80,
+				Protocol:        "http",
+			},
+			{
+				CertificateName: "cert",
+				ContainerPort:   4444,
+				HostPort:        443,
+				Protocol:        "https",
+			},
+		},
+		HealthCheck: models.HealthCheck{
+			Target:             "HTTPS:443/path/to/site",
+			Interval:           60,
+			Timeout:            60,
+			HealthyThreshold:   3,
+			UnhealthyThreshold: 3,
+		},
+	}
+
+	readSGHelper(mockAWS, "l0-test-env_id-env", "env_sg")
+	createSGHelper(t, mockAWS, "l0-test-lb_id-lb", "vpc_id")
+	readSGHelper(mockAWS, "l0-test-lb_id-lb", "lb_sg")
+
+	for _, port := range req.Ports {
+		ingressInput := &ec2.AuthorizeSecurityGroupIngressInput{}
+		ingressInput.SetGroupId("lb_sg")
+		ingressInput.SetCidrIp("0.0.0.0/0")
+		ingressInput.SetIpProtocol("TCP")
+		ingressInput.SetFromPort(int64(port.HostPort))
+		ingressInput.SetToPort(int64(port.HostPort))
+
+		mockAWS.EC2.EXPECT().
+			AuthorizeSecurityGroupIngress(ingressInput).
+			Return(&ec2.AuthorizeSecurityGroupIngressOutput{}, nil)
+	}
+
+	iamRoleInput := &iam.CreateRoleInput{}
+	iamRoleInput.SetRoleName("l0-test-lb_id-lb")
+	iamRoleInput.SetAssumeRolePolicyDocument(provider.DEFAULT_ASSUME_ROLE_POLICY)
+
+	mockAWS.IAM.EXPECT().
+		CreateRole(iamRoleInput).
+		Return(&iam.CreateRoleOutput{}, nil)
+
+	renderedLBRolePolicy, err := provider.RenderLoadBalancerRolePolicy(
+		"region",
+		"123456789012",
+		"l0-test-lb_id",
+		provider.DEFAULT_LB_ROLE_POLICY_TEMPLATE)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	putIAMPolicyInput := &iam.PutRolePolicyInput{}
+	putIAMPolicyInput.SetPolicyName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetRoleName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetPolicyDocument(renderedLBRolePolicy)
+
+	mockAWS.IAM.EXPECT().
+		PutRolePolicy(putIAMPolicyInput).
+		Return(&iam.PutRolePolicyOutput{}, nil)
+
+	listeners := make([]*elb.Listener, len(req.Ports))
+	for i, port := range req.Ports {
```

Comment:
I'm a little torn on this because it will make an already huge unit test even larger - but there's a lot of value in  avoiding logic like this to create our expected objects and instead build them by hand. For example, if someone made a change/mistake when setting up the `req` object above (e.g. remove `req.Ports` entirely), it could cause false positives.  

see bivers and my convo at https://github.com/quintilesims/layer0/pull/409/files/8c08f98d76bb3564ab00127c2b15db30bc0d67b0#diff-92d2c246bc12ec7b170dd0e92ef06122R92